src/ripple/basics/impl/Number.cpp

```diff
@@ -0,0 +1,732 @@
+    int
+    round() noexcept;
+};
+
+inline void
+Number::Guard::set_positive() noexcept
+{
+    sbit_ = 0;
+}
+
+inline void
+Number::Guard::set_negative() noexcept
+{
+    sbit_ = 1;
+}
+
+inline bool
+Number::Guard::is_negative() const noexcept
+{
+    return sbit_ == 1;
+}
+
+inline void
+Number::Guard::push(unsigned d) noexcept
+{
+    xbit_ = xbit_ || (digits_ & 0x0000'0000'0000'000F) != 0;
+    digits_ >>= 4;
+    digits_ |= (d & 0x0000'0000'0000'000FULL) << 60;
+}
+
+inline unsigned
+Number::Guard::pop() noexcept
+{
+    unsigned d = (digits_ & 0xF000'0000'0000'0000) >> 60;
+    digits_ <<= 4;
+    return d;
+}
+
+// Returns:
+//     -1 if Guard is less than half
+//      0 if Guard is exactly half
+//      1 if Guard is greater than half
+int
+Number::Guard::round() noexcept
+{
+    auto mode = Number::getround();
+
+    if (mode == towards_zero)
+        return -1;
+
+    if (mode == downward)
+    {
+        if (sbit_)
+        {
+            if (digits_ > 0 || xbit_)
+                return 1;
+        }
+        return -1;
+    }
+
+    if (mode == upward)
+    {
+        if (sbit_)
+            return -1;
+        if (digits_ > 0 || xbit_)
+            return 1;
+        return -1;
+    }
+
+    // assume round to nearest if mode is not one of the predefined values
+    if (digits_ > 0x5000'0000'0000'0000)
+        return 1;
+    if (digits_ < 0x5000'0000'0000'0000)
+        return -1;
+    if (xbit_)
+        return 1;
+    return 0;
+}
+
+// Number
+
+constexpr Number one{1000000000000000, -15, Number::unchecked{}};
+
+void
+Number::normalize()
+{
+    if (mantissa_ == 0)
+    {
+        *this = Number{};
+        return;
+    }
+    bool const negative = (mantissa_ < 0);
+    auto m = static_cast<std::make_unsigned_t<rep>>(mantissa_);
+    if (negative)
+        m = -m;
+    while ((m < minMantissa) && (exponent_ > minExponent))
+    {
+        m *= 10;
+        --exponent_;
+    }
+    Guard g;
+    while (m > maxMantissa)
+    {
+        if (exponent_ >= maxExponent)
+            throw std::overflow_error("Number::normalize 1");
+        g.push(m % 10);
+        m /= 10;
+        ++exponent_;
+    }
+    mantissa_ = m;
+    if ((exponent_ < minExponent) || (mantissa_ < minMantissa))
+    {
+        *this = Number{};
+        return;
+    }
+
+    auto r = g.round();
+    if (r == 1 || (r == 0 && (mantissa_ & 1) == 1))
+    {
+        ++mantissa_;
+        if (mantissa_ > maxMantissa)
+        {
+            mantissa_ /= 10;
+            ++exponent_;
+        }
+    }
+    if (exponent_ > maxExponent)
+        throw std::overflow_error("Number::normalize 2");
+
+    if (negative)
+        mantissa_ = -mantissa_;
+}
+
+Number&
+Number::operator+=(Number const& y)
+{
+    if (y == Number{})
+        return *this;
+    if (*this == Number{})
+    {
+        *this = y;
+        return *this;
+    }
+    if (*this == -y)
+    {
+        *this = Number{};
+        return *this;
+    }
+    assert(isnormal() && y.isnormal());
```

Comment:
How hard would it be to remove creating unnormalized numbers? If we can't, maybe we should change this assert so it works on normalized numbers if we ever hit this case. (here and other operators that assume normalized inputs)

Reply:
or maybe just `static_assert(!std::is_constant_evaluated() || isnormal());`?

Reply:
That's a possibility, but doing so would not eliminate the motivation for these asserts.